File: lib/src/main/java/org/mobiletoolkit/updater/VersionCheck.java
package org.mobiletoolkit.updater;

import android.support.annotation.NonNull;

import org.apache.maven.artifact.versioning.DefaultArtifactVersion;
import org.mobiletoolkit.updater.model.Version;
import org.mobiletoolkit.updater.model.VersionInfo;

/**
 * Created by Sebastian Owodzin on 15/05/2016.
 * Copyright © 2016 mobiletoolkit.org. All rights reserved.
 */
public class VersionCheck {

    private DefaultArtifactVersion appVersion;
    private String appApplicationId;

    private VersionInfo versionInfo;

    private Result result = null;

    public enum Result {
        UNSUPPORTED, OUTDATED, UP_TO_DATE
    }

    public VersionCheck(@NonNull String appVersionName, @NonNull String appApplicationId, @NonNull VersionInfo versionInfo) {
        this.appVersion = new DefaultArtifactVersion(appVersionName);
        this.appApplicationId = appApplicationId;
        this.versionInfo = versionInfo;
    }

    @NonNull
    public Result getResult() {
        if (null == result) {
            result = Result.OUTDATED;

            if (appApplicationId.equals(versionInfo.getLatestVersion().getApplicationId())) {
                if (0 != appVersion.compareTo(new DefaultArtifactVersion(versionInfo.getLatestVersion().getVersionName()))) {
                    result = Result.OUTDATED;
                } else {
                    result = Result.UP_TO_DATE;
                }
            }

            if (!Result.UP_TO_DATE.equals(result)) {
                for (Version version : versionInfo.getUnsupportedVersions()) {
                    if (appApplicationId.equals(version.getApplicationId())) {
                        int r = new DefaultArtifactVersion(version.getVersionName()).compareTo(appVersion);
                        if (0 <= new DefaultArtifactVersion(version.getVersionName()).compareTo(appVersion)) {
                            result = Result.UNSUPPORTED;
                            break;
                        }
                    }
                }
            }
        }

        return result;
    }

}

File: lib/src/main/java/org/mobiletoolkit/updater/VersionCheck.kt
package org.mobiletoolkit.updater

import org.apache.maven.artifact.versioning.DefaultArtifactVersion
import org.mobiletoolkit.updater.model.VersionsInfo

/**
 * Created by Sebastian Owodzin on 15/05/2016.
 * Copyright © 2017 mobiletoolkit.org. All rights reserved.
 */
public class VersionCheck(
        private val applicationId: String,
        private val versionName: String,
        val versionsInfo: VersionsInfo
) {
    public enum class Result {
        UP_TO_DATE, OUTDATED, UNSUPPORTED
    }
    public val result: Result
        get() {
            var resultValue = Result.OUTDATED

            val appVersion: DefaultArtifactVersion = DefaultArtifactVersion(versionName)

            if (applicationId == versionsInfo.latestVersionData.applicationId) {
                if (appVersion < DefaultArtifactVersion(versionsInfo.latestVersionData.versionName)) {
                    resultValue = Result.OUTDATED
                } else {
                    resultValue = Result.UP_TO_DATE
                }
            }

            if (resultValue != Result.UP_TO_DATE) {
                if (versionsInfo.unsupportedVersions.filter { it.applicationId == applicationId && DefaultArtifactVersion(it.versionName) >= appVersion }.isNotEmpty()) {
                    resultValue = Result.UNSUPPORTED
                }
            }

            return resultValue
        }
}
